File: cpp/bullcowHuman.cpp
#include <string>
#include <iostream>
#include <time.h>
#include <stdlib.h>

const int LEN = 4;

std::string generate()
{
    std::string res = "0123456789";
    for (int i = 0 ; i < LEN ; ++i)
    {
        int n = i + rand() % (10 - i);
        std::swap(res[i], res[n]);
    }
    res.resize(LEN);
    return res;
}

bool check(std::string s)
{
    for (int i = 0 ; i < LEN ; ++i)
    {
        if (s[i] < '0' || s[i] > '9'
            || s.find(s[i], i + 1) != std::string::npos)
        {
            return false;
        }
    }
    return true;
}

std::string inputGuess()
{
    std::string s;
    while (true)
    {
        std::cout << "Введите последовательность из 4 цифр: ";
        std::cin >> s;
        if (s.size() == LEN && check(s))
        {
            return s;
        }
    }
}

void countBullsAndCows(const std::string &secret, const std::string &guess)
{
    int bulls = 0;
    int cows = 0;
    for (int i = 0 ; i < LEN ; ++i)
    {
        if (guess[i] == secret[i])
        {
            ++bulls;
        }
        else if (secret.find(guess[i]) != std::string::npos)
        {
            ++cows;
        }
    }
    std::cout << "Быки: " << bulls << " Коровы: " << cows << "\n";
}

int main()
{
    srand((int)time(NULL));
    std::string secret = generate();
    std::string guess;
    while (true)
    {
        guess = inputGuess();
        if (guess == secret)
        {
            std::cout << "Вы выиграли!\n";
            break;
        }
        countBullsAndCows(secret, guess);
    }
}

File: python/bullcowHuman.py
#!/usr/bin/python3
import random

LEN = 4

def generate():
    return ''.join(random.sample('0123456789', LEN))

def check(s):
    for i in range(LEN):
        if s.find(s[i], i + 1) != -1:
            return False
    return True

def inputGuess():
    while True:
        s = input('Введите последовательность из 4 цифр: ')
        if len(s) == LEN and s.isdigit() and check(s):
            return s

def countBullsAndCows(secret, guess):
    bulls = 0
    cows = 0
    for i in range(LEN):
        if secret[i] == guess[i]:
            bulls += 1
        elif secret.find(guess[i]) != -1:
            cows += 1
    print(f'Быки: {bulls} Коровы: {cows}')

secret = generate()
while True:
    guess = inputGuess()
    if secret == guess:
        print('Вы выиграли!')
        break
    countBullsAndCows(secret, guess)
